### ISSUE TITLE
Add support for different folder names

### ISSUE THREAD

**Issue by DjKillerMemeStar** [erikbra/grate#198]
In a RoundhousE deployment I manage we use different names for the folders then those default specified. RoundhousE has support for these with these parameters "-up" "-sp" "-permissions". It would be nice if grate supported these, as we cannot update to grate without changing our internal workings with the tool.

The reason why we work like this is we have 3 folders

- "1. OnlyOnce" (-up)
- "2. OnChange" (-sp)
- "3. Everytime" (-permissions)

We like this way of working as this keeps the folder clutter down, and we rarely need to fiddle with permissions (we use AD mostly).

**Comment by wokket** [comment]
Hey mate, good catch 👍

I missed this when building the docco for RH features we don't yet support, so a PR to add custom directory names, or tweak the docco highlighting the missing feature is very welcome.

I'm hoping to get a new release published in the next day or two with 6 months of fixes/enhancements included, but this would be great to see in vNext+1 :)

**Comment by DjKillerMemeStar** [comment]
@wokket already on it :)

I would make it so we can add custom names for all folders if that is ok for you?

**Comment by wokket** [comment]
Yep. if we're going to do the work then full backwards compat with the options in RH is the goal.

**Comment by erikbra** [comment]
Great initiative, @DjKillerMemeStar . I didn't quite see the need for this feature when initially porting RH to grate, and thought it to be gold-plating, but of course, backwards-compatibility, and, as you say, maybe simplifying the directory structure, could be an aim here.

**Comment by DjKillerMemeStar** [comment]
@erikbra yes indeed! After my holiday i'll look into putting this in + including an option for the folder names roundhouse forgot/didn't want to include.

**Comment by erikbra** [comment]
Perfect, thanks a lot! 

And, if there are any questions I can help answering underway too, just give me a shout, of course.

**Comment by erikbra** [comment]
Just a thought, maybe we should just make the whole folder structure(s) configurable, just make it possible to make an ordered list of folder names, with folder type (OneTime, AnyTime, Everytime), so that it could fit different ways of thinking about this?

**Comment by erikbra** [comment]
Yes, this is a very good suggestion. But, I'm a bit sceptical to this example. Doesn't that mead that all folders of type "OneTimeScripts" would be bundled together?

https://docs.microsoft.com/en-us/dotnet/standard/commandline/syntax#multiple-arguments

So we cannot have an ordered list of folders, where different folder types are intertwined? We'd get all of one type, then the next, etc, wouldn't we?


I think, functionally, we need to support an ordered list of folders with no ordering of the type, e.g

setupstuff: Everytime
create-users: OneTime
upscripts: OneTime
storedprocs: AnyTime
test-data: OneTime
run-tests: Everytime

And expect them to be run in that order, with those type specifications.

**Comment by wokket** [comment]
Berko busy here so not a lot of thought gone into this, but assuming we can change the execution type of a file by filename (https://erikbra.github.io/grate/script-types/everytime/#everytime-scripts) then could we hit your "infinitely configurable folder options" by just allowing the original folder name suggestion, and combine that with custom file names?

In theory that would let you have a single folder (named what you like), running files in any order (via filename sorting) with any execution type (via filename content)??  If someone wants 2 or 3 or n folders after that they just need to override more folder paths via the original suggested change?

**Comment by erikbra** [comment]
I'm not sure I exactly understand your suggestion, however, I would very much like to be able supporting existing folder structures (say, for someone coming from other tools), and not force any changes to folder structure or script naming upon people.

Looking at GrateMigrator, most of the folders can just be put in an ordered list, and it can probably still work:

https://github.com/erikbra/grate/blob/main/grate/Migration/GrateMigrator.cs#L26

There are a few folders which are a bit "special", namely:

* `AlterDatabase`, which runs in a separate transactionscope/no transaction, and on the admin connection, 
* `BeforeMigration`, which runs in a separate transaction scope, on the default connection,
* `RunAfterCreateDatabase` (which is only run if the database was actually created, and 
* `Permissions` and `AfterMigration`, which are run in a separate transactionscope/no transaction, on the default connection.

So, if we need to make this fully configurable, we need to at lease support the following:

1. The list of folders should be ordered, independently of script type
2. Per folder, we need the following:
    a. Folder name
    b. Script types (AnyTime, OneTime, EveryTime)
    c. Whether to run within the transaction or not
    d. Which connection to run on (Should default to `ConnectionType.Default`, as per now, all folders run on default, except the `AlterDatabase` folder).

I think that should be the puzzle pieces we need. And then we need to abstract `KnownFolders` to be a `FolderConfiguration` or something, which provides an ordered list of folders to run, with the above config per folder.

**Comment by wokket** [comment]
Sorry, as I said a little scatter brained here atm.

Your latest post also renders my previous one redundant due to the extra complexities in meeting the 'Fully Configurable' spec.

My point was that your 2b above (we need to know what type of file it is) can already be catered for via special casing the filename (.everytime. etc).  If we allowed folder name overrides (the original request, meets your 2a) then we've opened up an awful lot of flexibility in file/folder layout.

**Comment by DjKillerMemeStar** [comment]
I've not had alot of time to think about this, but the best thing i can come up with is to have the ability to send over a json string in a parameter with all the info, this would make it pretty complex tho.

I'm also thinking my suggestion might work if we parse the parameters ourselfs instead of using the build in parameter parsers. But not sure how that works.

**Comment by erikbra** [comment]
Sorry for the lack of feedback here, I've been on vacation (sorry, but not sorry 😉 ). I have started to look at this, in a branch, but I see that some of the logic about stuff that should be ran in a separate, autonomous transaction, is not run in all cases. So I have wrote som (now failing) tests for that, and are working on solving those issues in the existing code base before changing it. I'll create a PR when I have had a bit of more time to look into it.

Having stumbled upon another existing code base (at work) with a bit of different folder naming conventions, I lean more and more towards being able to support _any_ folder structure, using a config (file, parameters, something), to be able to run on a collection of SQL scripts that have been run using another migration tool, with grate as a drop-in replacement, just configuring it to the code base/structure one is already using.
